emulator/app/src/i3c_socket.rs

```diff
@@ -133,26 +139,250 @@ fn handle_i3c_socket_connection(
+            Ok(()) => {
+                let outdata: OutgoingHeader = transmute!(out_header_bytes);
+                if target_addr != outdata.from_addr {
+                    return;
+                }
+                let resp_desc = outdata.response_descriptor;
+                let data_len = resp_desc.data_length() as usize;
+                let mut data = vec![0u8; data_len];
+
+                stream.set_nonblocking(false).unwrap();
+                stream
+                    .read_exact(&mut data)
+                    .expect("Failed to read message from socket");
+                stream.set_nonblocking(true).unwrap();
+
+                let pec = calculate_crc8((target_addr << 1) | 1, &data[..data.len() - 1]);
+                if pec == data[data.len() - 1] {
+                    self.check_response(&data[..data.len() - 1]);
+                } else {
+                    println!(
+                        "Received data with invalid CRC8: calclulated {:X} != received {:X}",
+                        pec,
+                        data[data.len() - 1]
+                    );
+                }
+
+                self.state = TestState::Finish;
+            }
+            Err(ref e) if e.kind() == ErrorKind::WouldBlock => {}
+            Err(e) => panic!("Error reading message from socket: {}", e),
+        }
+    }
+}
+
+struct TestRunner {
+    stream: TcpStream,
+    target_addr: u8,
+    passed: usize,
+    running: Arc<AtomicBool>,
+    tests: Vec<Test>,
+}
+
+impl TestRunner {
+    pub fn new(
+        stream: TcpStream,
+        target_addr: u8,
+        running: Arc<AtomicBool>,
+        tests: Vec<Test>,
+    ) -> Self {
+        Self {
+            stream,
+            target_addr,
+            passed: 0,
+            running,
+            tests,
+        }
+    }
+
+    pub fn run_tests(&mut self) {
+        for test in self.tests.iter_mut() {
+            test.run_test(self.running.clone(), &mut self.stream, self.target_addr);
+            if test.is_passed() {
+                self.passed += 1;
+            }
+        }
+        println!(
+            "Test Result: {} tests/{} total tests passed ",
+            self.passed,
+            self.tests.len()
+        );
+        self.running.store(false, Ordering::Relaxed);
+    }
+}
+
+fn prepare_private_write_cmd(to_addr: u8, data_len: u16) -> [u8; 9] {
+    let mut write_cmd = ReguDataTransferCommand::read_from_bytes(&[0; 8]).unwrap();
+    write_cmd.set_rnw(0);
+    write_cmd.set_data_length(data_len);
+
+    let cmd_words: [u32; 2] = transmute!(write_cmd);
+    let cmd_hdr = IncomingHeader {
+        to_addr,
+        command: cmd_words,
+    };
+    transmute!(cmd_hdr)
+}
+
+fn prepare_private_read_cmd(to_addr: u8) -> [u8; 9] {
+    let mut read_cmd = ReguDataTransferCommand::read_from_bytes(&[0; 8]).unwrap();
+    read_cmd.set_rnw(1);
+    read_cmd.set_data_length(0);
+    let cmd_words: [u32; 2] = transmute!(read_cmd);
```

Comment:
I'm a little worried about endianness biting us, especially in the (unlikely) event we ever run this code with a big endian host.

Could we add a test at least that checks that the packets generated have the bytes that we expect? (That way if something weird happens with endianness in the future, we'll at least have a test to let us know.)

Reply:
Good idea! Added tests for prepare_private_read and prepare_private_write functions.